streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java

```diff
@@ -291,14 +314,23 @@ private void putInternal(byte[] binaryKey, byte[] binaryValue) {
     }
 
     private Segment getSegment(long segmentId) {
-        int index = (int) (segmentId % segments.length);
+        if (segmentId > currentSegmentId - segments.length) {
+            int index = (int) (segmentId % segments.length);
 
-        if (segments[index] == null) {
-            segments[index] = new Segment(name + "-" + directorySuffix(segmentId), segmentId);
-            segments[index].openDB(context);
-        }
+            if (segments[index] != null && segments[index].id != segmentId) {
+                cleanup();
```

Comment:
Does this check & cleanup need to be here? Given that `cleanup()` is already invoked when you create new segments, when can this happen?